### PR TITLE
Feature: Added support for FTP_TLS (Weak Cryptography Issue)

### PR DESCRIPTION
**Pull request by fazledyn-or**
### Background 
I'm a Software Security Engineer from OpenRefactory-Inc. We are working with OpenSSF on [Project Alpha-Omega](https://alpha-omega.dev/). While triaging your project, our bug fixing tool generated the following message-

> In file: [ftp.py](https://github.com/jschneier/django-storages/blob/master/storages/backends/ftp.py#L87), method: _start_connection, a clear-text protocol such as FTP, Telnet or SMTP is used. These protocols [transfer data without any encryption,](https://cwe.mitre.org/data/definitions/319.html) which expose applications to a large range of risks. iCR suggested that data should be transferred over only secure transport channels.

As a result, I've added the support for `FTP_TLS` from `ftplib` library into the codebase. 

### Test Status
I copied the `test_ftp.py` file and modified it for `FTP_TLS`. Currently all tests are passing.

### CLA Requirements
This section is only relevant if your project requires contributors to sign a Contributor License Agreement (CLA) for external contributions. All contributed commits are already automatically signed off.

The meaning of a signoff depends on the project, but it typically certifies that committer has the rights to submit this work under the same license and agrees to a Developer Certificate of Origin (see [https://developercertificate.org/](https://developercertificate.org/) for more information).
- [Git Commit SignOff documentation](https://developercertificate.org/)

### Sponsorship and Support:

This work is done by the security researchers from OpenRefactory and is supported by the [Open Source Security Foundation (OpenSSF)](https://openssf.org/): [Project Alpha-Omega](https://alpha-omega.dev/). Alpha-Omega is a project partnering with open source software project maintainers to systematically find new, as-yet-undiscovered vulnerabilities in open source code - and get them fixed – to improve global software supply chain security.

The bug is found by running the Intelligent Code Repair (iCR) tool by OpenRefactory and then manually triaging the results.

### CODE REVIEW

**Comment by jschneier** [comment]
Hi and thanks for the PR, nice to see how minimal the amount of code needed to support this is.

Can you slim down the added tests to only verify the new parsing, setting of `secure`, and calling of `prot_p` i.e to only test the new additions.

File: storages/backends/ftp.py [hunk 1]
```diff
@@ -37,7 +39,7 @@ class FTPStorageException(Exception):
 class FTPStorage(Storage):
     """FTP Storage class for Django pluggable storage system."""
 
-    def __init__(self, location=None, base_url=None, encoding=None):
+    def __init__(self, location=None, base_url=None, encoding=None, secure=False):
```
Comment (jschneier) [inline storages/backends/ftp.py:42]: Please revert this change, see below.

Comment (jschneier) [inline storages/backends/ftp.py:42]: Please revert this change, see below.

File: storages/backends/ftp.py [hunk 2]
```diff
@@ -51,13 +53,14 @@ def __init__(self, location=None, base_url=None, encoding=None):
         self._config = self._decode_location(location)
         self._base_url = base_url
         self._connection = None
+        self._secure = secure
```
Comment (jschneier) [inline storages/backends/ftp.py:56]: Same here, see below.

File: storages/backends/ftp.py [hunk 3]
```diff
@@ -51,13 +53,14 @@ def __init__(self, location=None, base_url=None, encoding=None):
...
 
     def _decode_location(self, location):
         """Return splitted configuration data from location."""
         splitted_url = urlparse(location)
         config = {}
 
-        if splitted_url.scheme not in ("ftp", "aftp"):
+        if splitted_url.scheme not in ("ftp", "aftp", "ftps"):
```
Comment (jschneier) [inline storages/backends/ftp.py:63]: Please set `config[“secure”]` based on if the url is `ftps`.

File: storages/backends/ftp.py [hunk 4]
```diff
@@ -84,11 +87,13 @@ def _start_connection(self):
 
         # Real reconnect
         if self._connection is None:
-            ftp = ftplib.FTP()
+            ftp = ftplib.FTP_TLS() if self._secure else ftplib.FTP()
```
Comment (jschneier) [inline storages/backends/ftp.py:90]: And then branch the secure pieces based on the config value.

**Review by jschneier — CHANGES_REQUESTED** [review]
Thanks for the PR and nice to see how simple this nice addition is. I’ve commented on a better way I’d like to do this. As part of that change can you also slim down the test suite to just test these small additions, thanks!

File: storages/backends/ftp.py (hunk 3 above)
Comment (fazledyn-or) [reply storages/backends/ftp.py:63]: Thanks for the suggestions @jschneier.

I'll modify them add suggested and minimize the tests as instructed.

File: storages/backends/ftp.py (hunk 4 above)
Comment (fazledyn-or) [reply storages/backends/ftp.py]: Done. Check: https://github.com/jschneier/django-storages/pull/1320/commits/099196a250e943b9ec572b0d52373d4259314672

File: tests/test_ftp.py [hunk 5]
```diff
@@ -239,3 +244,49 @@ def test_close(self, mock_ftp, mock_storage):
...
+            "host": "localhost",
+            "user": "foo",
+            "active": False,
+            "path": "/",
+            "port": 2121,
+            "secure": True
+        }
+        self.assertEqual(config, wanted_config)
+
+    def test_decode_location_error(self):
```
Comment (jschneier) [inline tests/test_ftp.py:276]: You can remove this test.

File: tests/test_ftp.py [hunk 6]
```diff
@@ -239,3 +244,49 @@ def test_close(self, mock_ftp, mock_storage):
...
+            ftp.FTPStorage()
+
+    @patch("storages.backends.ftp.setting", return_value=URL_TLS)
+    def test_init_location_from_setting(self, mock_setting):
+        storage = ftp.FTPStorage()
+        self.assertTrue(mock_setting.called)
+        self.assertEqual(storage.location, URL_TLS)
+
+    def test_decode_location(self):
+        config = self.storage._decode_location(URL_TLS)
```
Comment (jschneier) [inline tests/test_ftp.py:264]: Given that `_decode_location` is called on `__init__` this test can be changed to

```python
def test_decode_location(self):
    wanted_config = {
        "passwd": "b@r",
        "host": "localhost",
        "user": "foo",
        "active": False,
        "path": "/",
        "port": 2121,
        "secure": True
    }
    self.assertEqual(self.storage._config, wanted_config)
```

File: tests/test_ftp.py [hunk 7]
```diff
@@ -239,3 +244,49 @@ def test_close(self, mock_ftp, mock_storage):
...
+
+class FTPTLSTest(TestCase):
+    def setUp(self):
+        self.storage = ftp.FTPStorage(location=URL_TLS)
+
+    def test_init_no_location(self):
+        with self.assertRaises(ImproperlyConfigured):
+            ftp.FTPStorage()
+
+    @patch("storages.backends.ftp.setting", return_value=URL_TLS)
```
Comment (jschneier) [inline tests/test_ftp.py:257]: Not sure this test adds anything

File: tests/test_ftp.py [hunk 8]
```diff
@@ -239,3 +244,49 @@ def test_close(self, mock_ftp, mock_storage):
         file_.is_dirty = True
         file_.read()
         file_.close()
+
+
+class FTPTLSTest(TestCase):
+    def setUp(self):
+        self.storage = ftp.FTPStorage(location=URL_TLS)
+
+    def test_init_no_location(self):
```
Comment (jschneier) [inline tests/test_ftp.py:253]: This test is already run in the base `FTPTest`, no need to duplicate.

**Review by jschneier — CHANGES_REQUESTED** [review]
Thanks for the updates it's looking good - just a few more changes to the tests needed and then I think we're good to merge.

File: tests/test_ftp.py [hunk 9]
```diff
@@ -239,3 +244,49 @@ def test_close(self, mock_ftp, mock_storage):
...
+        with self.assertRaises(ImproperlyConfigured):
+            self.storage._decode_location("foo")
+        with self.assertRaises(ImproperlyConfigured):
+            self.storage._decode_location("http://foo.pt")
+        # TODO: Cannot not provide a port
+        # with self.assertRaises(ImproperlyConfigured):
+        #     self.storage._decode_location('ftp://')
+
+    @patch("ftplib.FTP_TLS")
+    def test_start_connection(self, mock_ftp):
```
[the remark this answers is not in the document]
Comment (fazledyn-or) [reply tests/test_ftp.py:286]: Let me know if the following works - 

```py
    @patch("ftplib.FTP_TLS")
    def test_start_connection_calls_prot_p(self, mock_ftp):
        self.storage._start_connection()
        self.storage._connection.prot_p.assert_called_once()
```

I ran the test and it passed.

File: tests/test_ftp.py (hunk 8 above)
Comment (fazledyn-or) [reply tests/test_ftp.py:253]: Removed `test_init_no_location(self)`

File: tests/test_ftp.py (hunk 7 above)
Comment (fazledyn-or) [reply tests/test_ftp.py:257]: Check below

File: tests/test_ftp.py (hunk 6 above)
Comment (fazledyn-or) [reply tests/test_ftp.py:264]: Check below

File: tests/test_ftp.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (fazledyn-or) [reply tests/test_ftp.py:276]: I've removed the unnecessary test methods and this is what it looks like now - 
```py
class FTPTLSTest(TestCase):
    def setUp(self):
        self.storage = ftp.FTPStorage(location=URL_TLS)

    def test_decode_location(self):
        wanted_config = {
            "passwd": "b@r",
            "host": "localhost",
            "user": "foo",
            "active": False,
            "path": "/",
            "port": 2121,
            "secure": True
        }
        self.assertEqual(self.storage._config, wanted_config)

    @patch("ftplib.FTP_TLS")
    def test_start_connection(self, mock_ftp):
        self.storage._start_connection()
        self.assertIsNotNone(self.storage._connection)
        # Start active
        storage = ftp.FTPStorage(location=URL_TLS)
        storage._start_connection()

    @patch("ftplib.FTP_TLS")
    def test_start_connection_calls_prot_p(self, mock_ftp):
        self.storage._start_connection()
        self.storage._connection.prot_p.assert_called_once()

```
Please let me know if it needs any changes or not.

[the remark this answers is not in the document]
Comment (jschneier) [reply tests/test_ftp.py:276]: I think you can also remove `test_start_connection` unless I'm missing what that is testing. The rest looks good.

[the remark this answers is not in the document]
Comment (fazledyn-or) [reply tests/test_ftp.py]: You're correct. Somehow I missed it, thanks for notifying. I've updated the tests.

**Comment by jschneier** [comment]
@fazledyn-or just need to fix the lint issues then we are good to merge.

**Review by jschneier — APPROVED** [review]
